kubectl: fix deprecation warning bug

alexandercampbell opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://github.com/kubernetes/community/blob/master/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

**What this PR does / why we need it**:

Some kubectl commands were deprecated but would fail to print the
correct warning message when a flag was given before the command name.

	# Correctly prints the warning that "resize" is deprecated and
	# "scale" is now preferred.
	kubectl resize [...]

	# Should print the same warning but no warning is printed.
	kubectl --v=1 resize [...]

This was due to a fragile check on os.Args[1].

This commit implements a new function deprecatedCmd() that is used to
construct new "passthrough" commands which are marked as deprecated and
hidden.

Note that there is an existing "filters" system that may be preferable
to the system created in this commit. I'm not sure why the "filters"
array was not used for all deprecated commands in the first place.

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
NONE
```

Commit:
cmd: fix deprecation warning bug

Some kubectl commands were deprecated but would fail to print the
correct warning message when a flag was given before the command name.

	# Correctly prints the warning that "resize" is deprecated and
	# "scale" is now preferred.
	kubectl scale [...]

	# Should print the same warning but no warning is printed.
	kubectl --v=1 scale [...]

This was due to a fragile check on os.Args[1].

This commit implements a new function deprecatedCmd() that is used to
construct new "passthrough" commands which are marked as deprecated and
hidden.

Note that there is an existing "filters" system that may be preferable
to the system created in this commit. I'm not sure why the "filters"
array was not used for all deprecated commands in the first place.

alexandercampbell commented:
@k8s-ci-robot signed CLA

pwittrock commented:
clever

pwittrock commented:
@k8s-bot ok to test

pwittrock commented:
Looks good.  Are there existing tests for this?  If not, add some to https://github.com/kubernetes/kubernetes/blob/master/hack/make-rules/test-cmd-util.sh

alexandercampbell commented:
@pwittrock awesome, thanks. I was going to ask where tests for this should be written.

droot commented on pkg/kubectl/cmd/cmd.go:
```diff
@@ -284,6 +284,7 @@ func NewKubectlCommand(f cmdutil.Factory, in io.Reader, out, err io.Writer) *cob
 				NewCmdCreate(f, out, err),
 				NewCmdExposeService(f, out),
 				NewCmdRun(f, in, out, err),
+				deprecatedCmd("run-container", NewCmdRun(f, in, out, err)),
```
sweet :)

droot commented:
LGTM

Commit:
kubectl: improve docstring on deprecatedCmd

Commit:
test-cmd-util.sh: add test for deprecated commands

Commit:
Merge branch 'master' into correct-deprecation-errors

Commit:
.generated_docs: update to include "new" commands

droot commented on pkg/kubectl/cmd/cmd.go:
```diff
@@ -385,6 +391,20 @@ func printDeprecationWarning(command, alias string) {
+	cmd.Hidden = true
+	return cmd
+}
+
```
was thinking more about the unit test for this function. We can write a unit test which wraps a NoOp cobra.Command and captures the "deprecated" string in the output. Thoughts ? (happy to discuss in person)

alexandercampbell replied:
Sounds good to me. Will look at that.

alexandercampbell replied:
Unit test written. Thanks for the suggestion!

Commit:
kubectl: renmae deprecatedCmd -> deprecatedAlias

This will clear up some of the confusion around the deprecatedCmd /
Deprecated function.

Commit:
kubectl: make "Deprecated" a private function

There's no reason to export this function, so I've made it private.

Commit:
kubectl: write unit test for deprecatedAlias()

pwittrock commented:
@alexandercampbell Write a PTAL comment when you are ready to have us take another look

alexandercampbell commented:
Thanks! I will do that as soon as I figure out what's up with the CI.

Commit:
Merge branch 'master' into correct-deprecation-errors

Commit:
Merge branch 'master' into correct-deprecation-errors

droot commented on pkg/kubectl/cmd/cmd_test.go:
```diff
@@ -659,3 +662,38 @@ func genResponseWithJsonEncodedBody(bodyStruct interface{}) (*http.Response, err
+		t.Errorf("original command has name %q, expected %q",
+			original.Name(), "print")
+	}
+
```
hmm... I was thinking more along the lines of....invoking the deprecated command and then grabbing the output and check if deprecated warning is being emitted ?

pwittrock replied:
We could also include the output writer as a dependency defaulted to stdout, but overridden the test, and see what bytes that it writes.

alexandercampbell replied:
I've updated the test to include more coverage-- see 5e1a3fd.

droot replied:
Just looked at it that commit. Looks good to me!

droot reviewed (commented):
Took a quick look at the unit test and have a few comments.

Commit:
Merge branch 'master' into correct-deprecation-errors

alexandercampbell commented on pkg/kubectl/cmd/cmd_test.go:
```diff
@@ -25,9 +25,12 @@ import (
 	"net/http"
 	"os"
 	"reflect"
+	stdstrings "strings"
```
The name `strings` is taken by `k8s.io/kubernetes/pkg/util/strings` which is also imported in this package.

droot replied:
Got it. makes sense.

Commit:
kubectl: improve deprecatedAlias unit test

Two new behaviors are tested:

 1. The output message that deprecatedAlias gives when it is called must
    include the word "deprecatated" and the name of the new function
    that the user should use instead.
 2. The correct function must be called by the alias (alias should "fall
    back" to the functionality of the original.

droot reviewed (commented):
LGTM.

Commit:
Merge branch 'master' into correct-deprecation-errors

Commit:
test-cmd-util: fix deprecated commands test

Additionally, move the test down to ensure definition order matches run
order.

Commit:
docs: add new generated docs for deprecated commands

alexandercampbell commented:
I tentatively believe that the tests are green.
PTAL, and let me know if you guys want me to rebase out the extraneous commits.

pwittrock commented:
/lgtm

pwittrock commented:
/approve
